Library/Sender.ts

```diff
@@ -83,148 +92,185 @@ class Sender {
             this._enableDiskRetryMode = false;
             Logging.warn(Sender.TAG, "Ignoring request to enable disk retry mode. Sufficient file protection capabilities were not detected.")
         }
+        if (this._enableDiskRetryMode) {
+            // Starts file cleanup task
+            this._fileCleanupTimer = setTimeout(() => { this._fileCleanupTask(); }, Sender.CLEANUP_TIMEOUT);
+        }
+        else {
+            if (this._fileCleanupTimer) {
+                clearTimeout(this._fileCleanupTimer);
+            }
+        }
     }
 
-    public send(payload: Buffer, callback?: (v: string) => void) {
-        var endpointUrl = this._redirectedHost || this._config.endpointUrl;
+    public send(envelopes: Contracts.EnvelopeTelemetry[], callback?: (v: string) => void) {
+        if (envelopes) {
+            var endpointUrl = this._redirectedHost || this._config.endpointUrl;
 
-        // todo: investigate specifying an agent here: https://nodejs.org/api/http.html#http_class_http_agent
-        var options = {
-            method: "POST",
-            withCredentials: false,
-            headers: <{ [key: string]: string }>{
-                "Content-Type": "application/x-json-stream"
-            }
-        };
+            // todo: investigate specifying an agent here: https://nodejs.org/api/http.html#http_class_http_agent
+            var options = {
+                method: "POST",
+                withCredentials: false,
+                headers: <{ [key: string]: string }>{
+                    "Content-Type": "application/x-json-stream"
+                }
+            };
 
-        zlib.gzip(payload, (err, buffer) => {
-            var dataToSend = buffer;
-            if (err) {
-                Logging.warn(err);
-                dataToSend = payload; // something went wrong so send without gzip
-                options.headers["Content-Length"] = payload.length.toString();
-            } else {
-                options.headers["Content-Encoding"] = "gzip";
-                options.headers["Content-Length"] = buffer.length.toString();
-            }
+            let batch: string = "";
 
-            Logging.info(Sender.TAG, options);
+            envelopes.forEach(envelope => {
+                var payload: string = this._stringify(envelope);
+                if (typeof payload !== "string") {
+                    return;
```

Comment:
If there is a single "invalid" envelope that can't be stringified, (say in the middle of the payload), won't this case only half the batch to be sent?
Not that I can think of a normal case where JSON.stringify() would fail... Just looking for corner cases.

Reply:
We want to drop the individual "wrong" event and continue with others, I'm trying to keep previous logic, this change is mainly focused on having the envelopes available in this class instead of payload like before, allowing us to filter them for retriable response codes.